lib/model/rwfolder.go

```diff
@@ -1723,6 +1727,66 @@ func (f *sendReceiveFolder) IgnoresUpdated() {
 	f.IndexUpdated()
 }
 
+// deleteDir attempts to delete a directory. It checks for files/dirs inside
+// the directory and removes them if possible or returns an error if it fails
+func (f *sendReceiveFolder) deleteDir(dir string, ignores *ignore.Matcher, toBeScanned map[string]struct{}) error {
+	files, _ := f.fs.DirNames(dir)
+
+	toBeDeleted := make([]string, 0)
+
+	hasIgnored := false
+	hasKnown := false
+	hasToBeScanned := false
+
+	for _, dirFile := range files {
+		fullDirFile := filepath.Join(dir, dirFile)
+		if fs.IsTemporary(dirFile) || (ignores != nil && ignores.Match(fullDirFile).IsDeletable()) {
+			toBeDeleted = append(toBeDeleted, fullDirFile)
+		} else if ignores != nil && ignores.Match(fullDirFile).IsIgnored() {
+			hasIgnored = true
+		} else if cf, ok := f.model.CurrentFolderFile(f.ID, fullDirFile); !ok || cf.IsDeleted() || cf.IsInvalid() {
+			// Something appeared in the dir that we either are not
+			// aware of at all, that we think should be deleted or that
+			// is invalid, but not currently ignored -> schedule scan
+			toBeScanned[fullDirFile] = struct{}{}
+			hasToBeScanned = true
+		} else {
+			// Dir contains file that is valid according to db and
+			// not ignored -> something weird is going on
+			hasKnown = true
+		}
+	}
+
+	if hasToBeScanned {
+		return errDirHasToBeScanned
+	}
+	if hasIgnored {
+		return errDirHasIgnored
+	}
+	if hasKnown {
+		return errDirHasKnown
+	}
+
+	for _, del := range toBeDeleted {
```

Comment:
I think we could do this part before returning, to make some progress regardless?

Reply:
Maybe in the case where there is only something ignored in the dir that cannot be deleted, but in case there is something to be scanned or valid in there, I don't think we should delete anything, as we might not delete the dir in the end.